### ISSUE TITLE
Possible buffer overflow on heap in tjDecompressToYUV2

### ISSUE THREAD

**Issue by robertswiecki** [libjpeg-turbo/libjpeg-turbo#389]
Originaly reported in #388 

The picture at: https://user-images.githubusercontent.com/12155236/68846842-f95e7180-06cd-11ea-9207-80b34a91f939.jpg seems to trigger a buffer overflow on heap in tjDecompressToYUV2() even with the bd20344b0ae7d53e1e26e1c4339915032a8ea98f fix applied.

```shell
$ ./libjpeg-turbo/out/tjbench-static 68846842-f95e7180-06cd-11ea-9207-80b34a91f939.jpg -yuv

Testing YUV planar encoding/decoding

>>>>>  JPEG 4:4:4 --> BGR (Top-down)  <<<<<

Image size: 1002 x 891
=================================================================
==136873==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x62b000006080 at pc 0x0000004425fe bp 0x7ffd59651cb0 sp 0x7ffd59651460
READ of size 1002 at 0x62b000006080 thread T0
    #0 0x4425fd in __interceptor_memcpy.part.40 (/home/swiecki/fuzz/jpeg/libjpeg-turbo/out/tjbench-static+0x4425fd)
    #1 0x535324 in tjDecompressToYUVPlanes (/home/swiecki/fuzz/jpeg/libjpeg-turbo/out/tjbench-static+0x535324)
    #2 0x536609 in tjDecompressToYUV2 (/home/swiecki/fuzz/jpeg/libjpeg-turbo/out/tjbench-static+0x536609)
    #3 0x51f9e3 in decomp (/home/swiecki/fuzz/jpeg/libjpeg-turbo/out/tjbench-static+0x51f9e3)
    #4 0x51be2a in decompTest (/home/swiecki/fuzz/jpeg/libjpeg-turbo/out/tjbench-static+0x51be2a)
    #5 0x518ad3 in main (/home/swiecki/fuzz/jpeg/libjpeg-turbo/out/tjbench-static+0x518ad3)
    #6 0x7f04c5c6252a in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2352a)
    #7 0x424689 in _start (/home/swiecki/fuzz/jpeg/libjpeg-turbo/out/tjbench-static+0x424689)

0x62b000006080 is located 0 bytes to the right of 24192-byte region [0x62b000000200,0x62b000006080)
allocated by thread T0 here:
    #0 0x4e0bc0 in malloc (/home/swiecki/fuzz/jpeg/libjpeg-turbo/out/tjbench-static+0x4e0bc0)
    #1 0x5344d5 in tjDecompressToYUVPlanes (/home/swiecki/fuzz/jpeg/libjpeg-turbo/out/tjbench-static+0x5344d5)
    #2 0x536609 in tjDecompressToYUV2 (/home/swiecki/fuzz/jpeg/libjpeg-turbo/out/tjbench-static+0x536609)
    #3 0x51f9e3 in decomp (/home/swiecki/fuzz/jpeg/libjpeg-turbo/out/tjbench-static+0x51f9e3)
    #4 0x51be2a in decompTest (/home/swiecki/fuzz/jpeg/libjpeg-turbo/out/tjbench-static+0x51be2a)
    #5 0x518ad3 in main (/home/swiecki/fuzz/jpeg/libjpeg-turbo/out/tjbench-static+0x518ad3)
    #6 0x7f04c5c6252a in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2352a)

SUMMARY: AddressSanitizer: heap-buffer-overflow (/home/swiecki/fuzz/jpeg/libjpeg-turbo/out/tjbench-static+0x4425fd) in __interceptor_memcpy.part.40
Shadow bytes around the buggy address:
  0x0c567fff8bc0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c567fff8bd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c567fff8be0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c567fff8bf0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c567fff8c00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0c567fff8c10:[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c567fff8c20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c567fff8c30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c567fff8c40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c567fff8c50: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c567fff8c60: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
```

**Comment by robertswiecki** [comment]
You can also test it for crashes with e.g. electricFence or with libDuma

```shell
$ LD_PRELOAD=/usr/lib/libefence.so.0.0 ./libjpeg-turbo/out-none/tjbench-static ./68846842-f95e7180-06cd-11ea-9207-80b34a91f939.jpg -yuv

  Electric Fence 2.2 Copyright (C) 1987-1999 Bruce Perens <bruce@perens.com>

Testing YUV planar encoding/decoding

>>>>>  JPEG 4:4:4 --> BGR (Top-down)  <<<<<

Image size: 1002 x 891
Segmentation fault

$ gdb ./libjpeg-turbo/out-none/tjbench-static
..
(gdb) set environment LD_PRELOAD=/usr/lib/libefence.so.0.0
(gdb) r ./68846842-f95e7180-06cd-11ea-9207-80b34a91f939.jpg -yuv
Program received signal SIGSEGV, Segmentation fault.
__memmove_avx_unaligned_erms () at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:384
384	../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S: No such file or directory.
(gdb) bt
#0  __memmove_avx_unaligned_erms () at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:384
#1  0x0000555555567c37 in tjDecompressToYUVPlanes (handle=handle@entry=0x7ffff7ede928, jpegBuf=jpegBuf@entry=0x7ffff7ed2f48 "\377\330\377", <incomplete sequence \335>, jpegSize=jpegSize@entry=183, dstPlanes=dstPlanes@entry=0x7fffffffd490, width=<optimized out>, width@entry=1002, 
    strides=strides@entry=0x7fffffffd484, height=<optimized out>, flags=1024) at /home/swiecki/fuzz/jpeg/libjpeg-turbo/turbojpeg.c:1724
#2  0x0000555555568949 in tjDecompressToYUV2 (handle=0x7ffff7ede928, jpegBuf=0x7ffff7ed2f48 "\377\330\377", <incomplete sequence \335>, jpegSize=183, dstBuf=0x7ffff6d551b4 '\200' <repeats 200 times>..., width=1002, pad=1, height=891, flags=1024)
    at /home/swiecki/fuzz/jpeg/libjpeg-turbo/turbojpeg.c:1799
#3  0x000055555555cf0a in decomp (srcBuf=0x0, jpegBuf=0x7fffffffdaf0, jpegSize=0x7fffffffdaf8, dstBuf=<optimized out>, w=1002, h=891, subsamp=0, jpegQual=0, fileName=0x7fffffffdfc3 "./68846842-f95e7180-06cd-11ea-9207-80b34a91f939", tilew=1002, tileh=891)
    at /home/swiecki/fuzz/jpeg/libjpeg-turbo/tjbench.c:205
#4  0x000055555555b039 in decompTest (fileName=0x7fffffffdfc3 "./68846842-f95e7180-06cd-11ea-9207-80b34a91f939") at /home/swiecki/fuzz/jpeg/libjpeg-turbo/tjbench.c:712
#5  main (argc=<optimized out>, argv=<optimized out>) at /home/swiecki/fuzz/jpeg/libjpeg-turbo/tjbench.c:1003

(gdb) up
#1  0x0000555555567c37 in tjDecompressToYUVPlanes (handle=handle@entry=0x7ffff7ede928, jpegBuf=jpegBuf@entry=0x7ffff7ed2f48 "\377\330\377", <incomplete sequence \335>, jpegSize=jpegSize@entry=183, dstPlanes=dstPlanes@entry=0x7fffffffd490, width=<optimized out>, width@entry=1002, 
    strides=strides@entry=0x7fffffffd484, height=<optimized out>, flags=1024) at turbojpeg.c:1724
(gdb) info locals 
j = <optimized out>
yuvptr = {0x7ffff7ef9fc0, 0x7ffff7efbfc0, 0x7ffff7efdf40, 0x0, 0x7ffff6fe41b4, 0x7ffff7ee29a0, 0x7ffff7ede928, 0x555555589074 <consume_markers+692>, 0x120, 0x7ffff7edeb30}
crow = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0}
i = 2
sfi = 8
row = 0
retval = 0
jpegwidth = <optimized out>
jpegheight = <optimized out>
jpegSubsamp = <optimized out>
scaledw = <optimized out>
scaledh = <optimized out>
pw = {1002, 1002, 1002, 0, 33685504, 66561, 768, 1280, 0, 0}
ph = {891, 891, 891, 303104513, 16674, 0, 0, 0, 0, 0}
iw = {1008, 504, 504, 0, 0, 0, 0, 0, 0, 0}
tmpbufsize = <optimized out>
usetmpbuf = <optimized out>
th = {8, 8, 24, 0, 0, 0, 0, 0, 0, 0}
_tmpbuf = 0x7ffff7ef2180 '\200' <repeats 200 times>...
ptr = <optimized out>
outbuf = {0x7ffff7ee9428, 0x7ffff7eec428, 0x7ffff7eef428, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}
tmpbuf = {0x7ffff7ef9fc0, 0x7ffff7efbfc0, 0x7ffff7efdf40, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}
dctsize = 8
this = 0x7ffff7ede928
dinfo = 0x7ffff7edeb30

(gdb) up
#2  0x0000555555568949 in tjDecompressToYUV2 (handle=0x7ffff7ede928, jpegBuf=0x7ffff7ed2f48 "\377\330\377", <incomplete sequence \335>, jpegSize=183, dstBuf=0x7ffff6d551b4 '\200' <repeats 200 times>..., width=1002, pad=1, height=891, flags=1024 at turbojpeg.c:1799
(gdb) info locals 
dstPlanes = {0x7ffff6d551b4 '\200' <repeats 200 times>..., 0x7ffff6e2f122 "\377\377", 0x7ffff6f09090 '\200' <repeats 200 times>...}
pw0 = 1002
ph0 = 891
strides = {1002, 1002, 1002}
jpegSubsamp = <optimized out>
i = <optimized out>
jpegwidth = <optimized out>
jpegheight = <optimized out>
scaledw = <optimized out>
scaledh = <optimized out>
this = 0x7ffff7ede928
dinfo = 0x7ffff7edeb30

**Comment by Ch0ser** [comment]
Pardon me for asking that is this the cve-2019-2201? tks
sry I check both sites and figure out they are two things.

**Comment by dcommander** [comment]
I don't think so, but I am investigating.  CVE-2019-2201 was never reported upstream and describes a bug that was specific to the Neon assembly code.

**Comment by dcommander** [comment]
@Ch0ser I had to dig for the answer, but CVE-2019-2201 corresponds to #361, which was fixed in 2a9e3bd7430cfda1bc812d139e0609c6aca0b884 (libjpeg-turbo 2.0.3.)  The issue affected only the TurboJPEG API library, which is not as widely distributed or used as the libjpeg API library.  While that fix completely addressed the issue in the context of the TurboJPEG API library, the fix was incomplete in the context of the TJBench test program.  Thus, #388 was filed and fixed in c30b1e72dac76343ef9029833d1561de07d29bad (libjpeg-turbo 2.0.4.)  Since the latter commit affected only the TJBench test program, which isn't mission-critical, 2a9e3bd7430cfda1bc812d139e0609c6aca0b884 is the most relevant commit for the security issue.
